### ISSUE TITLE
InfoNCE d2q/symmetric with use_in_batch_negatives=false trains on a loss of exactly 0

### ISSUE THREAD

**Issue by kabirvashisht4-glitch** [NVIDIA-NeMo/Automodel#3869]
## Describe the bug

`infonce_loss` and `infonce_distill_loss` accept `direction="d2q"` (and
`"symmetric"`) together with `use_in_batch_negatives=False`. In that
configuration the document-to-query direction has **no candidates to score
against**, so its cross-entropy is exactly `0.0` with a zero gradient. Nothing
raises, nothing is logged — the run just trains on nothing.

Hard negatives are documents mined *per query*, so they are candidates for the
query-to-document direction only. `_one_direction` correctly passes `None` for
the reverse direction:

https://github.com/NVIDIA-NeMo/Automodel/blob/main/nemo_automodel/components/loss/infonce.py

```python
if direction == "q2d":
    return _one_direction(q, d, n)
if direction == "d2q":
    return _one_direction(d, q, None)          # <- negatives intentionally dropped
return 0.5 * (_one_direction(q, d, n) + _one_direction(d, q, None))
```

With `use_in_batch_negatives=False` that branch builds a single-column logit
tensor and a target of all zeros:

```python
pos = (query * doc).sum(dim=-1, keepdim=True)   # [B, 1]
logits = pos
target = torch.zeros(batch, device=device, dtype=torch.long)
...
return F.cross_entropy(logits, target)          # one class -> log(1) == 0
```

The existing guard only catches the case where *no* negatives were passed at
all, so it never fires here:

```python
if not use_in_batch_negatives and not has_hard_negs:
    raise ValueError("infonce_loss: no negatives provided")
```

`infonce_distill_loss` has the same shape of problem: the single-candidate
softmax makes both the student and the (detached) teacher distributions
degenerate, so KL/CE/MSE over them is identically zero.

## Steps/Code to reproduce bug

```python
import torch
from nemo_automodel.components.loss.infonce import infonce_loss

torch.manual_seed(0)
q = torch.randn(4, 8, requires_grad=True)
d = torch.randn(4, 8, requires_grad=True)
n = torch.randn(4, 3, 8, requires_grad=True)

for direction in ("q2d", "d2q", "symmetric"):
    for in_batch in (True, False):
        for t in (q, d, n):
            t.grad = None
        loss = infonce_loss(
            q, d, hard_negatives=n, direction=direction, use_in_batch_negatives=in_batch
        )
        loss.backward()
        dn = "None" if n.grad is None else f"{n.grad.abs().sum():.4f}"
        print(f"{direction:10s} in_batch={in_batch!s:5s} loss={loss.item():.6f} |dq|={q.grad.abs().sum():8.4f} |dn|={dn}")
```

```
q2d        in_batch=True  loss=8.986291 |dq|= 25.4025 |dn|=4.2906
q2d        in_batch=False loss=5.643632 |dq|= 17.9486 |dn|=20.5227
d2q        in_batch=True  loss=5.800508 |dq|= 20.0058 |dn|=None
d2q        in_batch=False loss=0.000000 |dq|=  0.0000 |dn|=None     <-- silent no-op
symmetric  in_batch=True  loss=7.393400 |dq|= 20.2241 |dn|=2.1453
symmetric  in_batch=False loss=2.821816 |dq|=  8.9743 |dn|=10.2613  <-- exactly q2d / 2
```

`2.821816 == 5.643632 / 2` exactly: the `d2q` half of `"symmetric"` contributes
literally nothing, so a config that asks for a symmetric objective silently gets
a halved one-directional one.

`infonce_distill_loss` behaves the same way (`loss=0.000000`, all student
gradients zero for `d2q` + `use_in_batch_negatives=False`).

## Why this is reachable from a config

`RetrievalDistillBiEncoderRecipe` builds both modules straight from YAML
(`self.cfg.get("infonce_loss")` / `self.cfg.get("infonce_distill_loss")`), and
`direction` and `use_in_batch_negatives` are plain top-level keys on that block —
see `examples/retrieval/distillation/nemotron3_embed_1b_distill.yaml`:

```yaml
infonce_distill_loss:
  _target_: nemo_automodel.components.loss.infonce.InfoNCEDistillLoss
  direction: q2d
  use_in_batch_negatives: true
```

The shipped example is safe, but flipping those two keys — a plausible thing to
try when a user has mined hard negatives and wants to drop in-batch negatives —
produces a run whose InfoNCE term is `0.0` at every step.

The recipe already guards a sibling case with exactly this failure mode:

```python
# Guard against a configured intermediate loss whose layer_pairs are empty
# while the weight is active: hooks would fire but the loss would silently
# short-circuit to zero (contributing no gradient).
```

## Expected behavior

The contradictory configuration should be rejected with a clear error rather
than silently producing a zero loss. `use_in_batch_negatives=False` needs
`hard_negatives` **and** `direction="q2d"`, because the document-to-query
direction can only draw candidates from the batch.

Reusing the hard negatives for `d2q` would not be a correct alternative: they
are document embeddings and cannot stand in for query-side candidates.

## Environment overview

- `main` @ `b1f4031d`
- CPU only, no GPU needed to reproduce.

## Additional context

Existing coverage in `tests/unit_tests/loss/test_infonce.py` asserts only
`torch.isfinite(loss)` for the direction matrix, and `0.0` is finite, so the
degenerate configurations pass today.
